fvm/environment/account-key-metadata/weight_and_revoked_group.go

```diff
@@ -0,0 +1,301 @@
+
+const (
+	maxRunLengthInWeightAndRevokedStatusGroup = math.MaxUint16
+
+	weightAndRevokedStatusSize      = 2
+	weightAndRevokedStatusIndex     = runLengthSize
+	weightAndRevokedStatusGroupSize = runLengthSize + weightAndRevokedStatusSize
+
+	revokedMask = 0x8000
+	weightMask  = 0x7fff
+)
+
+// getWeightAndRevokedStatus returns weight and revoked status of the given index from encoded data.
+// Received b is expected to only contain encoded weight and revoked status.
+func getWeightAndRevokedStatus(b []byte, index uint32) (bool, uint16, error) {
+	if len(b)%weightAndRevokedStatusGroupSize != 0 {
+		return false, 0, NewKeyMetadataMalfromedError(fmt.Sprintf("failed to get weight and revoked status: existing metadata is %d bytes, expect multiples of %d", len(b), weightAndRevokedStatusGroupSize))
+	}
+
+	for off := 0; off < len(b); off += weightAndRevokedStatusGroupSize {
+		runLength := parseRunLength(b, off)
+
+		if index < uint32(runLength) {
+			revoked, weight := parseWeightAndRevokedStatus(b, off+runLengthSize)
+			return revoked, weight, nil
+		}
+
+		index -= uint32(runLength)
+	}
+
+	return false, 0, NewKeyMetadataNotFoundError("failed to query weight and revoked status", index)
+}
+
+// appendWeightAndRevokedStatus appends given weight and revoked status to the given data.
+// New weight and revoked status can be appended in a new run-length group, or included in
+// last group by incrementing last group's run-length.
+// NOTE: this function can modify the given data.
+func appendWeightAndRevokedStatus(b []byte, revoked bool, weight uint16) (_ []byte, _ error) {
+	if len(b) == 0 {
+		return encodeWeightAndRevokedStatusGroup(1, revoked, weight), nil
+	}
+
+	if len(b)%weightAndRevokedStatusGroupSize != 0 {
+		return nil, NewKeyMetadataMalfromedError(fmt.Sprintf("failed to append weight and revoked status: existing metadata is %d bytes, expect multiples of %d", len(b), weightAndRevokedStatusGroupSize))
+	}
+
+	// Merge to last groupo
+	lastGroupOff := len(b) - weightAndRevokedStatusGroupSize
+	lastGroup := parseWeightAndRevokedStatusGroup(b, lastGroupOff)
+
+	if lastGroup.tryMerge(revoked, weight) {
+		b = append(b[:lastGroupOff], lastGroup.encode()...)
+		return b, nil
+	}
+
+	b = append(b, encodeWeightAndRevokedStatusGroup(1, revoked, weight)...)
+	return b, nil
+}
+
+// setRevokedStatus sets revoked status for the given index in the given data.
+// NOTE: this function can modify the given data.
+func setRevokedStatus(b []byte, index uint32) ([]byte, error) {
+	if len(b)%weightAndRevokedStatusGroupSize != 0 {
+		return nil, NewKeyMetadataMalfromedError(fmt.Sprintf("failed to set revoked status: existing metadata is %d bytes, expect multiples of %d", len(b), weightAndRevokedStatusGroupSize))
+	}
+
+	foundGroup := false
+	curOff := 0
+	for ; curOff < len(b); curOff += weightAndRevokedStatusGroupSize {
+		runLength := parseRunLength(b, curOff)
+
+		if index < uint32(runLength) {
```

Comment:
```suggestion
        // when this loop exits we are guaranteed that `index < max uint16` because `runLength` is uint16
		if index < uint32(runLength) {
```